### PR TITLE
XState clean up + v5

### PR DESCRIPTION
**Pull request by techsmyth**
Migrated to XState v5. 
Simplified logic surrounding our usage of XState.
Single definition of machine that is controlled by the server, no longer per instance. This makes management + updates massively simpler moving forward. 
Migration added to remove the machine definition column on Lifecycles. 

Strong typing now on the combination of machine def + actions / guards that are available. 

Todo:
- test suites need updating, think there are issues around invitations in particular

Note: despite the simpler + up to date setup, I was not yet able to resolve the exception within an action crashing the server, will need help on that. 

image

Later can also consider putting strong typing into the way we use XState, as that will again help track down issues significantly. 

## Summary by CodeRabbit

- **New Features**
	- Introduced `ApplicationLifecycleService` and `InvitationLifecycleService` to manage application and invitation lifecycles using state machines.
	- Added GraphQL resolvers for application and invitation lifecycles, enhancing lifecycle management capabilities.
	- Added a new method in `LookupResolverFields` to resolve user-related queries with authorization checks.
	- Introduced new methods in `RoleSetService` for accepting invitations and approving applications to role sets.

- **Bug Fixes**
	- Corrected import paths across various modules to ensure proper functionality.

- **Refactor**
	- Updated method names for consistency, including renaming state retrieval methods to `getLifecycleState`.
	- Simplified lifecycle management by removing old configuration objects and transitioning to a state machine approach.
	- Modified validation logic in `BaseHandler` to incorporate `InvitationEventInput`.

- **Documentation**
	- Enhanced comments and logging within methods for better clarity and traceability.

- **Tests**
	- Updated mock implementations to reflect method name changes in the `ApplicationService` and `InvitationService`.

### CODE REVIEW

**Comment by techsmyth** [comment]
@valentinyanakiev good if you can have a look at this one. It was a rabbit hole that I went into to try and track down the crashing server issues when exceptions are thrown in XState actions. Found the whole setup tough to work through, so gave our XState handling a big clean up. Very nice + much more maintainable result imho...

File: src/domain/access/role-set/role.set.resolver.mutations.ts [hunk 1]
```diff
@@ -673,47 +685,116 @@ export class RoleSetResolverMutations {
...
+      await this.lifecycleService.event({
+        machine: this.roleSetServiceLifecycleInvitation.getInvitationMachine(),
+        lifecycle: invitation.lifecycle,
+        eventName: InvitationLifecycleEvent.ACCEPTED,
+        agentInfo,
+        authorization: invitation.authorization,
+      });
+    }
+
+    return await this.invitationService.getInvitationOrFail(invitation.id);
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply src/domain/access/role-set/role.set.resolver.mutations.ts:798]: there won't be server crashes as there is a global exception handler

File: src/domain/access/application/application.service.lifecycle.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    new: {
+      on: {
+        APPROVE: {
+          guard: 'hasUpdatePrivilege',
+          target: ApplicationLifecycleState.APPROVING,
+        },
+        REJECT: {
+          guard: 'hasUpdatePrivilege',
+          target: ApplicationLifecycleState.REJECTED,
+        },
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply src/domain/access/application/application.service.lifecycle.ts]: fixed

File: src/domain/access/invitation/invitation.service.lifecycle.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    this.invitationMachineStatesOnly =
+      this.getInvitationLifecycleMachineWithOnlyStates();
+  }
+
+  public getState(lifecycle: ILifecycle): string {
+    return this.lifecycleService.getState(
+      lifecycle,
+      this.invitationMachineStatesOnly
+    );
+  }
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply src/domain/access/invitation/invitation.service.lifecycle.ts:21]: should be handled in https://app.zenhub.com/workspaces/alkemio-development-5ecb98b262ebd9f4aec4194c/issues/gh/alkem-io/server/4645

File: src/domain/access/invitation/invitation.service.lifecycle.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  public getNextEvents(lifecycle: ILifecycle): string[] {
+    return this.lifecycleService.getNextEvents(
+      lifecycle,
+      this.invitationMachineStatesOnly
+    );
+  }
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply src/domain/access/invitation/invitation.service.lifecycle.ts:28]: should be handled in https://app.zenhub.com/workspaces/alkemio-development-5ecb98b262ebd9f4aec4194c/issues/gh/alkem-io/server/4645

File: src/domain/access/application/application.service.lifecycle.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import { AnyStateMachine, createMachine } from 'xstate';
+
+@Injectable()
+export class ApplicationLifecycleService {
+  private applicationMachineStatesOnly: AnyStateMachine;
+
+  constructor(private lifecycleService: LifecycleService) {
+    this.applicationMachineStatesOnly =
+      this.getApplicationLifecycleMachineWithOnlyStates();
+  }
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply src/domain/access/application/application.service.lifecycle.ts:14]: To be handled in https://app.zenhub.com/workspaces/alkemio-development-5ecb98b262ebd9f4aec4194c/issues/gh/alkem-io/server/4645

**Review by valentinyanakiev — APPROVED** [review]

**Pull request merged**